src/Carousel.js

```diff
@@ -126,315 +134,350 @@ const defaultProps = {
-        triggerBrowserReflow(nextElement);
+    select(index, event);
+  };
 
-        this.safeSetState(
-          {
-            prevClasses: classNames('active', directionalClassName),
-            currentClasses: classNames(orderClassName, directionalClassName),
-          },
-          () =>
-            transitionEnd(nextElement, () => {
-              this.safeSetState(
-                { prevClasses: '', currentClasses: 'active' },
-                this.handleSlideEnd,
-              );
-              if (onSlideEnd) {
-                onSlideEnd();
-              }
-            }),
-        );
-      },
-    );
-  }
+  const handleSlideEnd = () => {
+    const pendingIndex = _pendingIndex.current;
+    _isSliding.current = false;
+    _pendingIndex.current = null;
 
-  componentWillUnmount() {
-    clearTimeout(this.timeout);
-    this.isUnmounted = true;
-  }
+    if (pendingIndex != null) to(pendingIndex);
+    else cycle();
+  };
+
+  const pause = () => {
+    _isPaused.current = true;
+    clearInterval(_interval.current);
+    _interval.current = null;
+  };
+
+  const handleMouseOut = () => {
+    cycle();
+  };
+
+  const handleMouseOver = () => {
+    if (pauseOnHover) pause();
+  };
 
-  handleTouchStart = e => {
-    this.setState({ touchStartX: e.changedTouches[0].screenX });
+  const handleTouchStart = e => {
+    touchStartX.current = e.changedTouches[0].screenX;
   };
 
-  handleTouchEnd = e => {
+  const handleTouchEnd = e => {
     // If the swipe is under the threshold, don't do anything.
     if (
-      Math.abs(e.changedTouches[0].screenX - this.state.touchStartX) <
+      Math.abs(e.changedTouches[0].screenX - touchStartX.current) <
       SWIPE_THRESHOLD
     )
       return;
 
-    if (e.changedTouches[0].screenX < this.state.touchStartX) {
+    if (e.changedTouches[0].screenX < touchStartX.current) {
       // Swiping left to navigate to next item.
-      this.handleNext(e);
+      handleNext(e);
     } else {
       // Swiping right to navigate to previous item.
-      this.handlePrev(e);
+      handlePrev(e);
     }
   };
 
-  handleSlideEnd = () => {
-    const pendingIndex = this._pendingIndex;
-    this._isSliding = false;
-    this._pendingIndex = null;
-
-    if (pendingIndex != null) this.to(pendingIndex);
-    else this.cycle();
-  };
-
-  handleMouseOut = () => {
-    this.cycle();
-  };
-
-  handleMouseOver = () => {
-    if (this.props.pauseOnHover) this.pause();
-  };
-
-  handleKeyDown = event => {
+  const handleKeyDown = event => {
     if (/input|textarea/i.test(event.target.tagName)) return;
 
     switch (event.key) {
       case 'ArrowLeft':
         event.preventDefault();
-        this.handlePrev(event);
+        handlePrev(event);
         break;
       case 'ArrowRight':
         event.preventDefault();
-        this.handleNext(event);
+        handleNext(event);
         break;
       default:
         break;
     }
   };
 
-  handleNextWhenVisible = () => {
+  useLayoutEffect(() => {
```

Comment:
we generally use `useIsomorphicEffect` instead of `useLayoutEffect` for SSR safety